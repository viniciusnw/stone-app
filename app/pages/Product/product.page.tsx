import React from "react";
import { View, Text, ActivityIndicator } from "react-native";
import { RouteProp, useNavigation, useRoute } from "@react-navigation/native";
import useSWRMutation from "swr/mutation";
import { Image } from "react-native";
import { Icon } from "@Components";
import { observer } from "mobx-react";
import { useStore as useCheckoutStore } from "@Contexts/checkout.context";
import { autorun } from "mobx";

const get = (url: string) => fetch(url).then((res) => res.json());

type ParamList = {
  Product: {
    productId: number;
  };
};

const Product: React.FC<any> = (props: any) => {
  const navigation = useNavigation();
  const route = useRoute<RouteProp<ParamList, "Product">>();
  const { cart } = useCheckoutStore();

  const {
    trigger,
    data: product,
    isMutating,
    error,
  } = useSWRMutation(
    `https://fakestoreapi.com/products/${route.params.productId}`,
    get
  );

  const currencyFormat = (num: number) => {
    return "R$" + num.toFixed(2).replace(/(\d)(?=(\d{3})+(?!\d))/g, "$1,");
  };

  const isAdded = (id: number) => !!cart.items.find((i) => i.id == id);

  const setNavigation = (p?: any) => {
    const _product = p;
    props.setState({
      bottomBar: {
        onPress: () =>
          isAdded(_product.id) ? cart.remove(_product.id) : cart.add(_product),
        text: isAdded(_product.id) ? "Remover" : "Adicionar",
        label: currencyFormat(_product.price),
        description: `ou 3x ${currencyFormat(_product.price / 3)}`,
      },
      topBar: {
        title: _product.title,
        back: navigation.goBack,
        cart: () => props.goTo("Cart"),
      },
    });
  };

  React.useEffect(
    () => autorun(() => product && setNavigation(product)),
    [cart.items, product]
  );

  React.useEffect(() => error && navigation.goBack(), [error]);

  React.useEffect(() => {
    if (route.params?.productId) trigger().then(p => setNavigation(p));
  }, []);

  return (
    <View style={{ flex: 1 }}>
      {!isMutating && !!product && (
        <>
          <Image
            style={{
              resizeMode: "contain",
              height: "40%",
              width: "100%",
              backgroundColor: "#fff",
            }}
            source={{
              uri: product.image,
            }}
          />

          <View style={{ padding: 25, gap: 10 }}>
            <Text
              style={{
                fontSize: 18,
                fontWeight: "bold",
              }}
            >
              {product.category}
            </Text>
            <Text>{product.description}</Text>

            <View style={{ flexDirection: "row", alignItems: "center" }}>
              <Rating rating={product.rating.rate} />
              <Text style={{ marginLeft: 10 }}>{product.rating.rate}</Text>
            </View>
          </View>
        </>
      )}

      {isMutating && (
        <View
          style={{ flex: 1, alignItems: "center", justifyContent: "center" }}
        >
          <ActivityIndicator size="large" color={"black"} />
        </View>
      )}
    </View>
  );
};

const Rating: React.FC<{ rating: number }> = ({ rating }) => {
  return [...Array(5)].map((el, i) =>
    i < rating && i + 1 > rating ? (
      <Icon key={i} icon={"starh"} />
    ) : i < rating ? (
      <Icon key={i} icon={"star"} />
    ) : (
      <Icon key={i} icon={"staro"} />
    )
  );
};

export default observer(Product);
