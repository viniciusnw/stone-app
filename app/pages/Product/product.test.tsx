import React from "react";
import renderer from "react-test-renderer";
import { lightTheme } from "@Theme";
import { ThemeProvider } from "styled-components/native";
import Product from "./product.page";
import "jest-styled-components";
import Wrapper from "@/app/navigation/wrapper.navigator";
import { mount, shallow} from 'enzyme';
import { mountToJson, shallowToJson } from 'enzyme-to-json';
import { Text } from "react-native";


jest.useFakeTimers({
  advanceTimers: 10000,
});

jest.mock("@react-navigation/native", () => {
  return {
    useNavigation: () => ({
      navigate: jest.fn(),
      dispatch: jest.fn(),
      goBack: jest.fn(),
    }),
    useRoute: () => ({
      params: {
        productId: 1,
      },
    }),
  };
});

jest.mock("swr/mutation", () => {
  return () => {
    return {
      trigger: () => {
        return {
          then: () => {
            return {
              image: require("@Assets/images/splash.png"),
              category: "category",
              description: "description",
              title: "title",
              price: 1000,
              rating: {
                rate: 5,
              },
            }
          }
        }
      },
      isMutating: false,
      data: {
        image: require("@Assets/images/splash.png"),
        category: "category",
        description: "description",
        title: "title",
        price: 1000,
        rating: {
          rate: 5,
        },
      },
    };
  };
});

test("ProductWrapper", () => {
  const wrapper = shallow(
    <ThemeProvider theme={lightTheme}>
      <Wrapper Page={Product} header bottomBarType={"action-bar"} />
    </ThemeProvider>
  )
  wrapper.update()
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});

test("Product", () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={lightTheme}>
        <Product setState={jest.fn} />
      </ThemeProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
