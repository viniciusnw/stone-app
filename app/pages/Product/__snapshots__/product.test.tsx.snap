// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <Image
    source={
      {
        "uri": 1,
      }
    }
    style={
      {
        "backgroundColor": "#fff",
        "height": "40%",
        "resizeMode": "contain",
        "width": "100%",
      }
    }
  />
  <View
    style={
      {
        "gap": 10,
        "padding": 25,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 18,
          "fontWeight": "bold",
        }
      }
    >
      category
    </Text>
    <Text>
      description
    </Text>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <Text />
      <Text />
      <Text />
      <Text />
      <Text />
      <Text
        style={
          {
            "marginLeft": 10,
          }
        }
      >
        5
      </Text>
    </View>
  </View>
</View>
`;

exports[`ProductWrapper 1`] = `
<ContextProvider
  value={
    {
      "color": {
        "black": "#111111",
        "border": "#ddd",
        "danger": "red",
        "default": "rgb(0,168,104)",
        "gray": "#666666",
        "grayDark": "#353535",
        "grayLight": "#8c8c8c",
        "primary": "rgb(10,245,61)",
        "success": "green",
        "warning": "orange",
        "white": "#ffffff",
      },
      "font": {
        "family": "'Helvetica Neue', Helvetica, sans-serif",
        "size": "14px",
      },
      "height": {
        "navbar": "62px",
      },
      "spacing": {
        "extraLarge": "40px",
        "extraSmall": "5px",
        "large": "30px",
        "medium": "20px",
        "regular": "15px",
        "small": "10px",
        "xxLarge": "60px",
      },
    }
  }
>
  <Memo(Wrapper)
    Page={
      {
        "$$typeof": Symbol(react.memo),
        "compare": null,
        "type": [Function],
      }
    }
    bottomBarType="action-bar"
    header={true}
  />
</ContextProvider>
`;
